pkg/config/application_config.go

```diff
@@ -54,3 +79,67 @@ func (a *ApplicationConfiguration) EqualsButServices(o *ApplicationConfiguration
 	}
 	return true
 }
+
+// AnnounceableAddress is a pair of node address in the form of "[host]:[port]"
+// with optional corresponding announced port to be used in version exchange.
+type AnnounceableAddress struct {
+	Address       string
+	AnnouncedPort uint16
+}
+
+// GetAddresses parses returns the list of AnnounceableAddress containing information
+// gathered from both deprecated Address / NodePort / AnnouncedNodePort and newly
+// created Addresses fields.
+func (a *ApplicationConfiguration) GetAddresses() ([]AnnounceableAddress, error) {
+	addrs := make([]AnnounceableAddress, 0, len(a.Addresses)+1)
+	if a.Address != nil || a.NodePort != nil || a.AnnouncedNodePort != nil {
+		var (
+			host     string
+			nodePort uint16
+		)
+		if a.Address != nil {
+			host = *a.Address
+		}
+		if a.NodePort != nil {
+			nodePort = *a.NodePort
+		}
+		addr := AnnounceableAddress{Address: net.JoinHostPort(host, strconv.Itoa(int(nodePort)))}
+		if a.AnnouncedNodePort != nil {
+			addr.AnnouncedPort = *a.AnnouncedNodePort
+		}
+		addrs = append(addrs, addr)
+	}
+	for i, addrStr := range a.Addresses {
+		if len(addrStr) == 0 {
+			return nil, fmt.Errorf("address #%d is empty", i)
+		}
+		var tail int
+		if hostEnd := strings.LastIndex(addrStr, "]"); hostEnd != -1 {
+			tail = hostEnd + 1
+		} else {
+			tail = strings.Index(addrStr, ":")
+			if tail == -1 {
+				return nil, fmt.Errorf("invalid format of address #%d: address:port:announcedPort form is expected, got %s", i, addrStr)
+			}
+		}
+		subs := strings.SplitN(addrStr[tail:], ":", 3) // [%zone]:port:announcedPort; host is trimmed
+		if len(subs) == 1 {
```

Comment:
Implemented, but https://github.com/nspcc-dev/neo-go/pull/2827#discussion_r1035707532, this algorithm doesn't work with (some of) IPv6 addresses.

Reply:
`[ipv6]:port` is always used for IPv6 to avoid ambiguity. Or you mean a plain `ipv6` with no ports?

Reply:
Yes, things like `3731:54:65fe::1:1` and `3731:54:65fe:2::a7` which should be treated as valid with zero `Port` and `AnnouncedPort`. Thus, I've adjusted this algorithm a bit so that it can handle these cases as far.